### PR TITLE
Create GKE cluster with custom labels

### PR DESCRIPTION
**Pull request by Laci21**
| Q               | A
| --------------- | ---
| Bug fix?        | no
| New feature?    | yes
| API breaks?     | no
| Deprecations?   | no
| Related tickets | partially #1579 
| License         | Apache 2.0

### What's in this PR?

- Make it possible to add custom labels for Kubernetes nodes from create GKE cluster API call
- Store the labels in DB
- Add a function to create foreign and referenced keys
- Add foreign keys for GKE specific DB tables
- Add the related SQL scripts
- Make sure delete from tables works with foreign keys
- Moved deletion of unused secrets from `DeleteFromDatabase()` as it is not a DB related method
- Modified Postman collection to show how to add labels from the create GKE cluster API
- Made possible to run `make generate-client` on Linux (earlier it failed for me)
- Added client generated code
- Modified test `TestCreateNodePoolsModelFromRequest` to use GKE labels

### Additional context

- Edge case: when a user tries to set a Pipeline specific label such as `nodepool.banzaicloud.io/name` or `node.banzaicloud.io/ondemand`. In this case they should recieve 400 response code as those labels are not allowed to be overwritten.

### Checklist

- [x] Implementation tested (with at least one cloud provider)
- [x] Error handling code meets the [guideline](https://github.com/banzaicloud/pipeline/blob/master/docs/error-handling-guide.md)
- [ ] Logging code meets the guideline (TODO)
- [x] OpenAPI and Postman files updated, client regenerated (`make generate-client`) (if needed)
- [ ] User guide and development docs updated (if needed)

### CODE REVIEW

File: cluster/gke_node_pools.go [hunk 1]
```diff
@@ -63,13 +71,19 @@ func createNodePoolsFromClusterModel(clusterModel *google.GKEClusterModel) ([]*g
 	for i := 0; i < nodePoolsCount; i++ {
 		nodePoolModel := clusterModel.NodePools[i]
 
+		labels := map[string]string{
+			pkgCommon.LabelKey:         nodePoolModel.Name,
+			pkgCommon.OnDemandLabelKey: strconv.FormatBool(!nodePoolModel.Preemptible),
+		}
+
+		for _, label := range nodePoolModel.Labels {
+			labels[label.Name] = label.Value
```
Comment (stoader) [inline cluster/gke_node_pools.go]: Return with error if user provided labels conflicts with Pipeline reserved labels

File: cluster/manager_delete.go [hunk 2]
```diff
@@ -247,6 +248,17 @@ func deleteDnsRecordsOwnedByCluster(cluster CommonCluster) error {
 	return nil
 }
 
+func deleteUnusedSecrets(cluster CommonCluster) error {
```
Comment (stoader) [inline cluster/manager_delete.go]: Make this a method of `Manager` and use the logger that `Manager` holds.

File: cluster/manager_delete.go [hunk 3]
```diff
@@ -352,6 +364,11 @@ func (m *Manager) deleteCluster(ctx context.Context, cluster CommonCluster, forc
 	// delete from proxy from kubeProxyCache if any
 	m.DeleteKubeProxy(cluster)
 
+	err = deleteUnusedSecrets(cluster)
```
Comment (stoader) [inline cluster/manager_delete.go]: This supersedes `ClusterModel.preDelete` thus `ClusterModel.preDelete` can be deleted.

File: cluster/gke_node_pools.go (hunk 1 above)
Comment (Laci21) [reply cluster/gke_node_pools.go]: Isn't it enough to check earlier in the `CreateClusterGKE.Validate()` function, [here](https://github.com/Laci21/pipeline/blob/create-gke-cluster-with-custom-labels/pkg/cluster/gke/gke.go#L119)?
(I'll modify that implementation to check for some kind of Banzai Cloud or Pipeline specific label name, instead of checking all labels one by one, once that pattern is decided.)

File: cluster/manager_delete.go (hunk 2 above)
Comment (Laci21) [reply cluster/manager_delete.go]: How about the solution I added in [this](https://github.com/banzaicloud/pipeline/pull/1638/commits/c7c2fb5c11448722be0adb857cee563aeea69603) commit? This way, this function uses the logger of `Manager` (as it gets it as a parameter), but I also kept the way how similar functions are implemented in this file, like e.g. [DeleteAllResources()](https://github.com/banzaicloud/pipeline/blob/c7c2fb5c11448722be0adb857cee563aeea69603/cluster/manager_delete.go#L72) or[ deleteDnsRecordsOwnedByCluster()](https://github.com/banzaicloud/pipeline/blob/c7c2fb5c11448722be0adb857cee563aeea69603/cluster/manager_delete.go#L233), so to keep it as a simple function in this file and not a method of `Manager`. Do you think this is reasonable like this?

File: cluster/manager_delete.go (hunk 3 above)
Comment (Laci21) [reply cluster/manager_delete.go]: I think I have already deleted it in [this](https://github.com/banzaicloud/pipeline/pull/1638/commits/cb32a0b6f7feb5d35cba22d9e31e216312083adb) commit.

File: Makefile [hunk 4]
```diff
@@ -95,7 +95,7 @@ endif
 
 .PHONY: docker-build
 docker-build: ## Builds go binary in docker image
-	docker run -it -v $(PWD):/go/src/${PACKAGE} -w /go/src/${PACKAGE} golang:${GOLANG_VERSION}-alpine go build -o pipeline_linux ${BUILD_PACKAGE}
+	docker run -it -v $$(PWD):/go/src/${PACKAGE} -w /go/src/${PACKAGE} golang:${GOLANG_VERSION}-alpine go build -o pipeline_linux ${BUILD_PACKAGE}
```
Comment (sagikazarmark) [inline Makefile]:
```suggestion
	docker run -it -v $${PWD}:/go/src/${PACKAGE} -w /go/src/${PACKAGE} golang:${GOLANG_VERSION}-alpine go build -o pipeline_linux ${BUILD_PACKAGE}
```

File: internal/providers/google/gke_cluster_model.go [hunk 5]
```diff
@@ -22,6 +22,13 @@ import (
 	"github.com/jinzhu/gorm"
 )
 
+// TableName constants
+const (
+	GKEClusterModelTableName  = "google_gke_clusters"
```
Comment (sagikazarmark) [inline internal/providers/google/gke_cluster_model.go]: If not used outside of this package, these constants should not be exported.

File: internal/providers/google/model.go [hunk 6]
```diff
@@ -42,5 +44,25 @@ func Migrate(db *gorm.DB, logger logrus.FieldLogger) error {
 		"table_names": strings.TrimSpace(tableNames),
 	}).Info("migrating provider tables")
 
-	return db.AutoMigrate(tables...).Error
+	err := db.AutoMigrate(tables...).Error
+	if err != nil {
+		return err
+	}
+
+	err = model.AddForeignKey(db, logger, &model.ClusterModel{}, &GKEClusterModel{}, "ClusterID")
```
Comment (sagikazarmark) [inline internal/providers/google/model.go]: IIRC foreign keys can be defined on structs as tags. Although visually not really appealing, it might be better to have all table definitions in a single place (the model).

File: pkg/errors/errors.go [hunk 7]
```diff
@@ -26,40 +26,41 @@ var (
...
+	ErrorClusterNotReady                       = errors.New("Cluster not ready yet")
+	ErrorNilCluster                            = errors.New("<nil> cluster")
+	ErrorWrongKubernetesVersion                = errors.New("Wrong kubernetes version for master/nodes. The required minimum kubernetes version is 1.8.x ")
+	ErrorDifferentKubernetesVersion            = errors.New("Different kubernetes version for master and nodes")
+	ErrorLocationEmpty                         = errors.New("Location field is empty")
+	ErrorRequiredLocation                      = errors.New("location is required")
+	ErrorRequiredSecretId                      = errors.New("Secret id is required")
+	ErrorCloudInfoK8SNotSupported              = errors.New("Not supported key in case of amazon")
+	ErrorNodePoolNotProvided                   = errors.New("At least one 'nodepool' is required for creating or updating a cluster")
+	ErrorNodePoolLabelClashesWithPipelineLabel = errors.New("NodePool label cannot be set to Pipeline specific label")
```
Comment (sagikazarmark) [inline pkg/errors/errors.go]:
```suggestion
	ErrorNodePoolLabelClashesWithPipelineLabel = errors.New("NodePool label conflicts with Pipeline reserved label name")
```

File: internal/providers/google/model.go (hunk 6 above)
Comment (stoader) [reply internal/providers/google/model.go]: GORM doesn't create FKs based on struct tags, it uses the "foreighkey" struct tag solely for tracking the dependencies between models.

File: Makefile (hunk 4 above)
Comment (Laci21) [reply Makefile]: Ah, I missed that, thank you!

File: internal/providers/google/gke_cluster_model.go (hunk 5 above)
Comment (Laci21) [reply internal/providers/google/gke_cluster_model.go]: Understood and fixed. Need to get to used to this naming convention :)

File: internal/providers/google/model.go (hunk 6 above)
Comment (sagikazarmark) [reply internal/providers/google/model.go]: Sad :(
